File: Algorithms/python/binarySearch.py
# Implementacão Imperativa
def RecBinarySearch(li,target):
    def helper(li,min_index,max_index,target):
        if min_index > max_index:
            return False  
        mid = (min_index + max_index) // 2  # Calculate midpoint index once

        if li[mid] == target:
            return True
        elif li[mid] > target:
            return helper(li, min_index, mid - 1, target)
        else:
            return helper(li, mid + 1, max_index, target)
    return helper(li, 0, len(li) - 1, target)
        
        


# Testing setup - - - - - - - - - - - - - - - - - - - - -
test_list = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17]

print(RecBinarySearch(test_list, 12))
# - - - - - - - - - - - - - - - - - - - - - -

File: Algorithms/scala/binarySearch.scala
object Script{
    def RecBinarySearch(li: List[Int], target : Int):Boolean  = {
        def helper(l: Int, r: Int): Boolean = {
            if (l > r) false
            val mid = (l + r) / 2
            if (li(mid) == target) true
            else if (li(mid) < target) helper(mid + 1, r)
            else helper(l, mid - 1)
    }
        return helper(0,li.length -1)
    }
    def main(args: Array[String]): Unit = {
        val li = List(1,2,3,4,5)
        val target = 7
        val result = RecBinarySearch(li, target)
        println(result)
    }
}